pvlib/ivtools.py

```diff
@@ -0,0 +1,334 @@
+    if v_oc is None:
+        v_oc = v[-1]
+    if i_sc is None:
+        i_sc = i[0]
+    if mp is not None:
+        v_mp, i_mp = mp
+    else:
+        v_mp, i_mp = _find_mp(v, i)
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(v, i, vlim, v_oc)
+
+    if not np.isnan(beta0):
+        beta3, beta4 = _find_beta3_beta4(v, i, beta0, beta1, ilim, i_sc)
+
+    # calculate single diode parameters from regression coefficients
+    result = _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp,
+                                       v_oc)
+
+    if result is None:
+        raise RuntimeError("Parameter extraction failed. Try increasing the"
+                           "number of data points in the IV curve")
+    else:
+        IL, I0, Rsh, Rs, nNsVth = result
+        return IL, I0, Rsh, Rs, nNsVth
+
+
+def _find_mp(v, i):
+    """
+    Finds voltage and current at maximum power point.
+
+    Parameters
+    ----------
+    v : ndarray
+        1D array containing voltage at each point on the IV curve, increasing
+        from 0 to v_oc inclusive, of `float` type [V]
+
+    i : ndarray
+        1D array containing current at each point on the IV curve, decreasing
+        from i_sc to 0 inclusive, of `float` type [A]
+
+    Returns
+    -------
+    v, i : tuple
+        voltage ``v`` and current ``i`` at the maximum power point [V], [A]
+    """
+    p = v * i
+    idx = np.argmax(p)
+    return v[idx], i[idx]
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
+
+
+def _find_beta0_beta1(v, i, vlim, v_oc):
+    # Get intercept and slope of linear portion of IV curve.
+    # Start with V =< vlim * v_oc, extend by adding points until slope is
+    # negative (downward).
+    beta0 = np.nan
+    beta1 = np.nan
+    idx = np.searchsorted(v, vlim * v_oc)
+    while idx <= len(v):
+        coef = np.polyfit(v[:idx], i[:idx], deg=1)
+        if coef[0] < 0:
+            # intercept term
+            beta0 = coef[1].item()
+            # sign change of slope to get positive parameter value
+            beta1 = -coef[0].item()
+            break
+        else:
+            idx += 1
+    return beta0, beta1
+
+
+def _find_beta3_beta4(v, i, beta0, beta1, ilim, i_sc):
+    # Subtract the IV curve from the linear fit.
+    y = beta0 - beta1 * v - i
+    x = np.array([np.ones_like(v), v, i]).T
+    # Select points where y > ilim * i_sc to regress log(y) onto x
+    result = np.linalg.lstsq(x[y > ilim * i_sc], np.log(y[y > ilim * i_sc]),
+                             rcond=None)
+    coef = result[0]
+    beta3 = coef[1].item()
+    beta4 = coef[2].item()
+    return beta3, beta4
+
+
+def _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp, v_oc):
+    failed = False
+    if any(np.isnan([beta0, beta1, beta3, beta4])):
+        failed = True
```

Comment:
Zen of python says explicit is better.

Here and below, and then above in the calling function, imo use exceptions and propagate them up.

- remove `failed = False`
- instead of `failed = False`, `raise RuntimeError("beta coefficient was NaN")` 
- remove `else` and dedent, because next block can't be reached if no error
- on line 322, check `I0`
       if (I0_vmp > 0) and (I0_voc > 0):
           raise RuntimeError("I0 was less than zero")
- remove `if failed: results = None` and remove `else`, then just `return IL, I0, Rsh, Rs, nNsVth`
- finally in `fit_sde_sandia` on line 240 instead of `if result is None:` either just remove that section and do nothing, because you are already raising an exception, or place a try above and put except there instead, to modify the message.

Reply:
I'm all for improving the code for others to understand. One challenge here is that `np.linalg.lstsq` and `np.polyfit` throw an unhelpful "does not converge" error if there is any NaN in the data.  I want to avoid exiting with that error while assigning np.nan to SDE parameters that aren't determined.